File: trab3/ruby/pprint.rb
# encoding: utf-8

def format_num num
  #"% 12.8f" % num rescue "%s" % num
  num.to_s
end

def get_max_width column
  column.collect{|ele| format_num(ele).length}.max
end

def print_table table, header=true
  return if table.empty?
  pads = []
  (0..(table[0].length-1)).each do |i|
    pads << get_max_width(table.collect{|row| row[i]})
  end
  puts "┌─#{pads.map{|pad| '─' * pad}.join('─┬─')}─┐"
  rowify = proc do |row|
    frow = row.zip(pads).map do |line, pad|
      format_num(line).rjust(pad)
    end
    puts "│ #{frow.join(' │ ')} │"
  end
  if header
    rowify.call table[0]
    puts "├─#{pads.map{|pad| '─' * pad}.join('─┼─')}─┤"
    table[1..-1].each &rowify
  else
    table.each &rowify
  end
  puts "└─#{pads.map{|pad| '─' * pad}.join('─┴─')}─┘"
end

File: trab3/python/pprint.py
# coding=utf-8
"""Pretty prints."""
import sys


def format_num(num):
    """Format a number according to given places.
    Adds commas, etc."""

    try:
        #return "{0: .8f}".format(float(num))
        return str(num)

    except (ValueError, TypeError):
        return "{0:<}".format(num)


def get_max_width(table, index):
    """Get the maximum width of the given column index"""
    return max([len(format_num(row[index])) for row in table])


def print_table(table, out=sys.stdout, realtable=False):
    """Prints out a table of data, padded for alignment
    @param out: Output stream (file-like object)
    @param table: The table to print. A list of lists.
    Each row must have the same number of columns. """
    col_paddings = []

    if not table:
        return

    for i in range(len(table[0])):
        col_paddings.append(get_max_width(table, i))

    print >> out, '┌─' + '─┬─'.join(map(lambda c: '─' * int(c), col_paddings)) + '─┐'
    print >> out, '│ ' + ' │ '.join(map(lambda (k, c): format_num(c).rjust(col_paddings[k]), enumerate(table[0]))) + ' │'
    for row in table[1:]:
        if realtable:
            print >> out, '├─' + '─┼─'.join(map(lambda c: '─' * int(c), col_paddings)) + '─┤'
        print >> out, '│ ' + ' │ '.join(map(lambda (k, c): format_num(c).rjust(col_paddings[k]), enumerate(row))) + ' │'
    print >> out, '└─' + '─┴─'.join(map(lambda c: '─' * int(c), col_paddings)) + '─┘'

